src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -186,11 +188,49 @@ private static MemoryAddress bufferCopy(MemoryAddress dest, MemorySegment buffer
         return dest;
     }
 
-    public static void checkFunctionTypes(MethodType mt, FunctionDescriptor cDesc) {
-        if (mt.parameterCount() != cDesc.argumentLayouts().size())
-            throw new IllegalArgumentException("arity must match!");
-        if ((mt.returnType() == void.class) == cDesc.returnLayout().isPresent())
-            throw new IllegalArgumentException("return type presence must match!");
+    private static boolean isPrimitiveSupported(Class<?> carrier) {
+        return carrier == byte.class
+            || carrier == short.class
+            || carrier == char.class
+            || carrier == int.class
+            || carrier == long.class
+            || carrier == float.class
+            || carrier == double.class;
+    }
+
+    private static void checkCompatibleType(Class<?> carrier, MemoryLayout layout, long addressSize) {
```

Comment:
I recalled that we do this check in the foreign memory access API (see `LayoutPath.java`):

```
if (!carrier.isPrimitive() || carrier == void.class || carrier == boolean.class // illegal carrier?
                || Wrapper.forPrimitiveType(carrier).bitWidth() != layout.bitSize()) { // carrier has the right size?
            throw new IllegalArgumentException("Invalid carrier: " + carrier + ", for layout " + layout);
        }
```
Can we perhaps unify a little? It seems like your check should subsume the existing check (at least for ValueLayouts).

Reply:
I've moved some of the common code into jdk.internal.foreign.Utils